Source/Serilog.Exceptions/Reflection/ReflectionInfoExtractor.cs

```diff
@@ -0,0 +1,104 @@
+namespace Serilog.Exceptions.Reflection
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Linq.Expressions;
+    using System.Reflection;
+
+    /// <summary>
+    /// Utility class that analyzes type using reflection and provides
+    /// information about properties to be used in destructuring process.
+    /// </summary>
+    internal class ReflectionInfoExtractor
+    {
+        private readonly object lockObj = new();
+        private readonly Dictionary<Type, ReflectionInfo> reflectionInfoCache = new();
+        private readonly IList<PropertyInfo> baseExceptionPropertiesForDestructuring;
+
+        public ReflectionInfoExtractor() => this.baseExceptionPropertiesForDestructuring = GetExceptionPropertiesForDestructuring(typeof(Exception));
+
+
+        public ReflectionInfo GetOrCreateReflectionInfo(Type valueType)
+        {
+            lock (this.lockObj)
+            {
+                if (!this.reflectionInfoCache.TryGetValue(valueType, out var reflectionInfo))
+                {
+                    reflectionInfo = this.GenerateReflectionInfoForType(valueType);
+                    this.reflectionInfoCache.Add(valueType, reflectionInfo);
+                }
+
+                return reflectionInfo;
+            }
+        }
+
+        private static Func<object, object> GenerateFastGetterForProperty(Type type, PropertyInfo property)
```

Comment:
Going forward, we might be able to leverage source generators for this, given that we know most exception types ahead of time.

Reply:
I get the idea, thanks.